### ISSUE TITLE
[external bug report][from claude] Non-square matmul output corrupts CB memory with small buffer_factor

### ISSUE THREAD

**Issue by zoecarver** [tenstorrent/tt-lang#365]
**Note: matmul has not landed on main yet. This issue is filed against PR #299 (Add matmul op).**

## Description

A non-square matmul `A[1,4] @ B[4,4] -> C[1,4]` followed by an elementwise accumulate produces incorrect results when the matmul destination CBs have small `buffer_factor` values. Increasing `buffer_factor` to large values (>=20) works around the issue.

**Conditions:**
- Matmul with non-square output (e.g., `[1,4]` output from `[1,4] @ [4,4]`)
- Result used in a subsequent elementwise operation (e.g., `acc + mm`)
- Small `buffer_factor` on matmul destination CBs (e.g., 3-5)

**Key observation from generated C++:** The matmul codegen loops iterate `max(input_dims)` for all output axes. For `[1,4] @ [4,4] -> [1,4]`, the outer loop runs `j < 4` instead of `j < 1`, writing 16 tiles via `pack_tile` into a CB that only reserved 4. The extra 12 tiles write past the reserved region, likely corrupting adjacent CB memory.

## Reproducer

Three tests isolate the bug:
- **Test A:** matmul in a loop, no accumulate, small bf -> **PASS** (matmul alone is fine)
- **Test B:** matmul + elementwise accumulate, small bf -> **FAIL** (corruption)
- **Test C:** matmul + elementwise accumulate, large bf -> **PASS** (workaround)

```python
"""
BUG: Non-square matmul [1,4] @ [4,4] -> [1,4] with small buffer_factor
     corrupts results when combined with elementwise accumulate.

- Test A: matmul in loop, no accumulate, small bf -> PASS
- Test B: matmul + accumulate, small bf -> FAIL
- Test C: matmul + accumulate, large bf -> PASS (workaround)

Only the combination of non-square matmul output + elementwise add + small
buffer_factor triggers the bug. Large buffer_factor works around it.
"""
import torch
import ttnn
import ttl

D = 4
K_BATCH = 4


# Test A: just matmul in a loop, no accumulate, keep last result only
@ttl.kernel(grid=(1, 1))
def test_matmul_loop_only(A, B, out):
    a_cb = ttl.make_circular_buffer_like(A, shape=(1, D), buffer_factor=1)
    b_cb = ttl.make_circular_buffer_like(B, shape=(K_BATCH, D), buffer_factor=2)
    mm_cb = ttl.make_circular_buffer_like(A, shape=(1, K_BATCH), buffer_factor=3)
    out_cb = ttl.make_circular_buffer_like(A, shape=(1, K_BATCH), buffer_factor=4)

    @ttl.compute()
    def compute():
        with a_cb.wait() as av:
            for _ in range(8):
                with b_cb.wait() as bv, mm_cb.reserve() as mm:
                    mm.store(ttl.math.matmul(av, bv, mm))
                with mm_cb.wait() as mv, out_cb.reserve() as o:
                    o.store(mv)

    @ttl.datamovement()
    def dm_read():
        with a_cb.reserve() as blk:
            tx = ttl.copy(A[0:1, 0:D], blk); tx.wait()
        for ki in range(8):
            s = ki * K_BATCH
            with b_cb.reserve() as blk:
                tx = ttl.copy(B[s:s+K_BATCH, 0:D], blk); tx.wait()

    @ttl.datamovement()
    def dm_write():
        for _ in range(8):
            with out_cb.wait() as blk:
                tx = ttl.copy(blk, out[0:1, 0:K_BATCH]); tx.wait()


# Test B: matmul + accumulate, NUM_K=2 (1 init + 1 add), small bf
@ttl.kernel(grid=(1, 1))
def test_matmul_acc_2(A, B, out):
    a_cb = ttl.make_circular_buffer_like(A, shape=(1, D), buffer_factor=1)
    b_cb = ttl.make_circular_buffer_like(B, shape=(K_BATCH, D), buffer_factor=2)
    mm_cb = ttl.make_circular_buffer_like(A, shape=(1, K_BATCH), buffer_factor=3)
    acc_cb = ttl.make_circular_buffer_like(A, shape=(1, K_BATCH), buffer_factor=4)
    out_cb = ttl.make_circular_buffer_like(A, shape=(1, K_BATCH), buffer_factor=5)

    @ttl.compute()
    def compute():
        with a_cb.wait() as av:
            with b_cb.wait() as bv, acc_cb.reserve() as ac:
                ac.store(ttl.math.matmul(av, bv, ac))
            with b_cb.wait() as bv, mm_cb.reserve() as mm:
                mm.store(ttl.math.matmul(av, bv, mm))
            with mm_cb.wait() as mv, acc_cb.wait() as old:
                with acc_cb.reserve() as new:
                    new.store(old + mv)
        with acc_cb.wait() as final, out_cb.reserve() as o:
            o.store(final)

    @ttl.datamovement()
    def dm_read():
        with a_cb.reserve() as blk:
            tx = ttl.copy(A[0:1, 0:D], blk); tx.wait()
        for ki in range(2):
            s = ki * K_BATCH
            with b_cb.reserve() as blk:
                tx = ttl.copy(B[s:s+K_BATCH, 0:D], blk); tx.wait()

    @ttl.datamovement()
    def dm_write():
        with out_cb.wait() as blk:
            tx = ttl.copy(blk, out[0:1, 0:K_BATCH]); tx.wait()


# Test C: matmul + accumulate, NUM_K=4, LARGE buffer_factor (workaround)
@ttl.kernel(grid=(1, 1))
def test_matmul_acc_4(A, B, out):
    a_cb = ttl.make_circular_buffer_like(A, shape=(1, D), buffer_factor=1)
    b_cb = ttl.make_circular_buffer_like(B, shape=(K_BATCH, D), buffer_factor=2)
    mm_cb = ttl.make_circular_buffer_like(A, shape=(1, K_BATCH), buffer_factor=20)
    acc_cb = ttl.make_circular_buffer_like(A, shape=(1, K_BATCH), buffer_factor=21)
    out_cb = ttl.make_circular_buffer_like(A, shape=(1, K_BATCH), buffer_factor=22)

    @ttl.compute()
    def compute():
        with a_cb.wait() as av:
            with b_cb.wait() as bv, acc_cb.reserve() as ac:
                ac.store(ttl.math.matmul(av, bv, ac))
            for _ in range(3):
                with b_cb.wait() as bv, mm_cb.reserve() as mm:
                    mm.store(ttl.math.matmul(av, bv, mm))
                with mm_cb.wait() as mv, acc_cb.wait() as old:
                    with acc_cb.reserve() as new:
                        new.store(old + mv)
        with acc_cb.wait() as final, out_cb.reserve() as o:
            o.store(final)

    @ttl.datamovement()
    def dm_read():
        with a_cb.reserve() as blk:
            tx = ttl.copy(A[0:1, 0:D], blk); tx.wait()
        for ki in range(4):
            s = ki * K_BATCH
            with b_cb.reserve() as blk:
                tx = ttl.copy(B[s:s+K_BATCH, 0:D], blk); tx.wait()

    @ttl.datamovement()
    def dm_write():
        with out_cb.wait() as blk:
            tx = ttl.copy(blk, out[0:1, 0:K_BATCH]); tx.wait()


if __name__ == "__main__":
    device = ttnn.open_device(device_id=0)
    def to_dram(t):
        return ttnn.from_torch(t, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
                               device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG)

    torch.manual_seed(42)
    a = torch.randn(32, D * 32, dtype=torch.bfloat16)
    b = torch.randn(8 * K_BATCH * 32, D * 32, dtype=torch.bfloat16)

    def ref(num_k):
        acc = torch.zeros(32, K_BATCH * 32)
        for ki in range(num_k):
            s = ki * K_BATCH * 32
            acc += a.float() @ b[s:s+K_BATCH*32].float()
        return acc.bfloat16().float()

    # Test A: matmul loop only (no accumulate), check last result
    print("=== A: matmul in loop, no accumulate (8 iters, keep last) ===")
    last_expected = a.float() @ b[7*K_BATCH*32:8*K_BATCH*32].float()
    last_expected = last_expected.bfloat16().float()
    out_a = to_dram(torch.zeros(32, K_BATCH * 32, dtype=torch.bfloat16))
    test_matmul_loop_only(to_dram(a), to_dram(b), out_a)
    ra = ttnn.to_torch(out_a).float()
    pcc = torch.corrcoef(torch.stack([ra.flatten(), last_expected.flatten()]))[0, 1].item()
    print(f"  PCC={pcc:.6f}  {'PASS' if pcc > 0.99 else 'FAIL'}")

    # Test B: matmul + acc, NUM_K=2, small bf
    print("\n=== B: matmul + accumulate, NUM_K=2, small bf ===")
    exp2 = ref(2)
    out_b = to_dram(torch.zeros(32, K_BATCH * 32, dtype=torch.bfloat16))
    test_matmul_acc_2(to_dram(a), to_dram(b), out_b)
    rb = ttnn.to_torch(out_b).float()
    pcc = torch.corrcoef(torch.stack([rb.flatten(), exp2.flatten()]))[0, 1].item()
    print(f"  PCC={pcc:.6f}  {'PASS' if pcc > 0.99 else 'FAIL'}")

    # Test C: matmul + acc, NUM_K=4, large bf (workaround)
    print("\n=== C: matmul + accumulate, NUM_K=4, large bf (workaround) ===")
    exp4 = ref(4)
    out_c = to_dram(torch.zeros(32, K_BATCH * 32, dtype=torch.bfloat16))
    test_matmul_acc_4(to_dram(a), to_dram(b), out_c)
    rc = ttnn.to_torch(out_c).float()
    pcc = torch.corrcoef(torch.stack([rc.flatten(), exp4.flatten()]))[0, 1].item()
    print(f"  PCC={pcc:.6f}  {'PASS' if pcc > 0.99 else 'FAIL'}")

    ttnn.close_device(device)
```

## HW Output

```
=== A: matmul in loop, no accumulate (8 iters, keep last) ===
  PCC=0.999979  PASS

=== B: matmul + accumulate, NUM_K=2, small bf ===
  PCC=0.492948  FAIL

=== C: matmul + accumulate, NUM_K=4, large bf (workaround) ===
  PCC=0.999975  PASS
```

## Analysis

The generated C++ for the matmul shows:
```cpp
cb_reserve_back(mm_cb, 4);  // reserve 4 tiles (correct for [1,4] output)
for (j = 0; j < 4; j++) {   // BUG: should be j < 1 for 1 output row
  for (k = 0; k < 4; k++) {
    for (l = 0; l < 4; l++) {
      matmul_tiles(a_cb, b_cb, ...);
    }
    pack_tile(0, mm_cb, j*4+k);  // writes offsets 0..15, not 0..3!
  }
}
cb_push_back(mm_cb, 4);  // push 4 tiles
```

The outer loop uses `max(input_dims)=4` instead of the actual output row count of 1. This writes 16 tiles into a CB that reserved 4, overwriting 12 tiles past the reserved region.

**Workaround:** Use `buffer_factor >= 20` for any CB that is a matmul destination with non-square output shape. This provides enough L1 padding to absorb the overwrite.

## Environment

- Runner: tt-lang hardware (Blackhole)
- Compiler: tt-lang latest
- Simulator: not tested (may detect as deadlock due to stricter CB checking)
